control-plane/csi-controller/src/controller.rs

```diff
@@ -0,0 +1,726 @@
+
+        let volume = rpc::csi::Volume {
+            capacity_bytes: size as i64,
+            volume_id: volume_uuid,
+            volume_context: args.parameters.clone(),
+            content_source: None,
+            accessible_topology: vt_mapper.volume_accessible_topology(pinned_volume),
+        };
+
+        debug!("Created volume: {:?}", volume);
+
+        Ok(Response::new(CreateVolumeResponse {
+            volume: Some(volume),
+        }))
+    }
+
+    async fn delete_volume(
+        &self,
+        request: tonic::Request<DeleteVolumeRequest>,
+    ) -> Result<tonic::Response<DeleteVolumeResponse>, tonic::Status> {
+        let args = request.into_inner();
+
+        debug!("Request to delete volume: {:?}", args);
+        MayastorApiClient::get_client()
+            .delete_volume(&args.volume_id)
+            .await
+            .map_err(|e| {
+                Status::internal(format!(
+                    "Failed to delete volume {}, error = {:?}",
+                    args.volume_id, e
+                ))
+            })?;
+
+        Ok(Response::new(DeleteVolumeResponse {}))
+    }
+
+    async fn controller_publish_volume(
+        &self,
+        request: tonic::Request<ControllerPublishVolumeRequest>,
+    ) -> Result<tonic::Response<ControllerPublishVolumeResponse>, tonic::Status> {
+        let args = request.into_inner();
+
+        debug!("Request to publish volume: {:?}", args);
+        if args.readonly {
+            return Err(Status::invalid_argument(
+                "Read-only volumes are not supported",
+            ));
+        }
+
+        let protocol = match args.volume_context.get("protocol") {
+            Some(p) => parse_protocol(p)?,
+            None => {
+                return Err(Status::invalid_argument(
+                    "No protocol specified for publish volume request",
+                ))
+            }
+        };
+
+        if args.node_id.is_empty() {
+            return Err(Status::invalid_argument("Node ID must not be empty"));
+        }
+
+        if args.volume_id.is_empty() {
+            return Err(Status::invalid_argument("Volume ID must not be empty"));
+        }
+
+        match args.volume_capability {
+            Some(c) => {
+                check_volume_capabilities(&[c])?;
+            }
+            None => {
+                return Err(Status::invalid_argument("Missing volume capability"));
+            }
+        };
+
+        let node_id = normalize_hostname(args.node_id);
+        let volume_id = args.volume_id.to_string();
+
+        // Check if the volume is already published.
+        let volume = MayastorApiClient::get_client()
+            .get_volume(&volume_id)
+            .await?;
+        let uri = if let Some(state) = volume.state.as_ref() {
+            let curr_proto = state.protocol.to_string();
+
+            // Volume is aready published, make sure the protocol matches and get URI.
+            if curr_proto != "none" {
```

Comment:
I wonder if we should check for `volume.child` here, which is what signifies that the volume is published. The protocol should actually be part of that.